[MRG] Bug fix and new feature: fix implementation of average precision score and add eleven-point interpolated option (7356 rebased)

GaelVaroquaux opened the pull request:
Rebased version of #7356

Fixes #4577 and #6377

What does this implement/fix? Explain your changes.

This adds an optional interpolation parameter to both average_precision_score. By default, the value is set to None which replicates the existing behavior, but there is also a 'eleven_point' option that implements the strategy described in Stanford's Introduction to Information Retrieval.



<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue
<!-- Example: Fixes #1234 -->


#### What does this implement/fix? Explain your changes.


#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

ndingwall committed:
Adds support for step-wise interpolation to auc and average_precision_score

GaelVaroquaux committed:
DOC: Simpler precision-recall example, remove 11pt

Remove the eleven average precision score

Add better tests.

GaelVaroquaux commented:
I need to address the pending comments of #7356 and make the example pretty again, as it was rendered a bit horrible by the merge to master

GaelVaroquaux commented:
I reordered a bit the example (no major change though). I find it more readable on the html output.

GaelVaroquaux commented:
After a lot of back and worth, I have finally convinced myself that I trust this implementation of average_precision_score. I was worried about dealing with duplicates. So I added a stringent test.

This is ready for merge.

Can I have a review, @agramfort, @vene, @amueller

amueller commented on doc/whats_new.rst:
```diff
@@ -6,6 +6,35 @@ Release history
 ===============
 
 Version 0.19
+Version 0.18.2
```
rebase issue you need to move this to the stuff below

amueller commented on examples/model_selection/plot_precision_recall.py:
```diff
@@ -54,11 +54,32 @@
+possible with a recall value at least equal to the target value.
+The most common choice is 'eleven point' interpolated precision, where the
+desired recall values are [0, 0.1, 0.2, ..., 1.0]. This is the metric used in
+`The PASCAL Visual Object Classes (VOC) Challenge <http://citeseerx.ist.psu.edu
```
Maybe this metrics was? sometimes? Or just remove that sentence?

amueller commented on sklearn/metrics/ranking.py:
```diff
@@ -149,14 +149,34 @@ def average_precision_score(y_true, y_score, average="macro",
+            For each of the recall values, r, in {0, 0.1, 0.2, ..., 1.0},
+            compute the arithmetic mean of the first precision value with a
+            corresponding recall >= r. This is the metric used in the Pascal
+            Visual Objects Classes (VOC) Challenge and is as described in the
```
Maybe not mention Pascal VOC?

GaelVaroquaux replied:
Well, they describe it in the docs, though don't use it in the code. Maybe we should mention the docs.

amueller replied:
yeah  I don't know... but the current text could be misleading.

amueller commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -122,6 +128,37 @@ def _average_precision(y_true, y_score):
+
+def _interpolated_average_precision_slow(y_true, y_score):
+    """A second implementation for the eleven-point interpolated average
+    precision used by Pascal VOC. This should produce identical results to
```
descripted in ir-book?

amueller commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -470,19 +507,24 @@ def test_precision_recall_curve_pos_label():
+    precision_recall_auc = _average_precision_slow(y_true, probas_pred)
+    interpolated_average_precision = _interpolated_average_precision_slow(
+        y_true, probas_pred)
+    assert_array_almost_equal(precision_recall_auc, 0.859, 3)
```
where does that number come from

GaelVaroquaux replied:
Agreed. Ping @ndingwall: we don't like to have hard-coded numbers in our test suites without being able to rederive them.

amueller commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -470,19 +507,24 @@ def test_precision_recall_curve_pos_label():
                                               np.zeros_like(probas_pred))
     precision_recall_auc = auc(r, p)
-    assert_array_almost_equal(precision_recall_auc, 0.75, 3)
+    assert_array_almost_equal(precision_recall_auc, 0.75)
```
where does that number come from?

GaelVaroquaux replied:
More rebase errors :).

Will do

GaelVaroquaux replied:
This number is actually meaningless, as the way above to compute the AUC of the PR curve is not the right way: it does not cater for edge effects. On a constant prediction, the average precision score is the TPR, here .5.

I am removing these lines. Anyhow, I added a correct test (which passes) at a different location.

amueller commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -510,15 +552,15 @@ def test_precision_recall_curve_toydata():
         assert_array_almost_equal(p, [0.5, 0., 1.])
         assert_array_almost_equal(r, [1., 0.,  0.])
-        assert_almost_equal(auc_prc, 0.25)
+        assert_almost_equal(auc_prc, 0.5)
```
maybe add a short explanation? Why is this one right and the other one wasn't?

amueller replied:
explanation still missing ;)

amueller commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -567,31 +609,47 @@ def test_precision_recall_curve_toydata():
+
+def test_average_precision_constant_values():
+    # Check the average_precision_score of a constant predictor is
+    # the tps
```
tps? tpr?

amueller replied:
not addressed

amueller commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -567,31 +609,47 @@ def test_precision_recall_curve_toydata():
+    y_true[::4] = 1
+    # And a constant score
+    y_score = np.ones(100)
+    # The precision is then the fraction of positive whatever the recall
```
"for all thresholds and recall values"?

GaelVaroquaux replied:
No, because there is only one threshold.

amueller commented on doc/whats_new.rst:
```diff
@@ -6,6 +6,35 @@ Release history
+............
+
+   - Added a `'eleven-point'` interpolated average precision option to
+     :func:`metrics.ranking.average_precision_score` as used in the `PASCAL
```
is it?

amueller reviewed (commented):
LGTM, gonna check out the example now

amueller commented on examples/model_selection/plot_precision_recall.py:
```diff
@@ -71,76 +92,100 @@
+plt.title('Precision-Recall example: AUC={0:0.2f}'.format(
+        average_precision["micro"]))
 plt.legend(loc="lower left")
 plt.show()
```
I'd remove the show so both pop up at once.

amueller commented on examples/model_selection/plot_precision_recall.py:
```diff
@@ -71,76 +92,100 @@
+precision["micro"], recall["micro"], _ = precision_recall_curve(Y_test.ravel(),
     y_score.ravel())
-average_precision["micro"] = average_precision_score(y_test, y_score,
+average_precision["micro"] = average_precision_score(Y_test, y_score,
```
Is this standard? I'm not sure I understand what that means. Maybe do binary for this example?

amueller replied:
So I think multi-label is not that common, and I'm not sure this is a common strategy for multi-class. as a matter of fact, we don't have multi-class average precision (there is a PR for multi-class AUC and there is a bunch of literature on that!).
So I vote going binary here.

GaelVaroquaux replied:
I'll do a binary first, and then illustrate micro

vene replied:
I have certainly seen this before in NLP even for multiclass.

amueller commented on sklearn/metrics/ranking.py:
```diff
@@ -110,10 +106,14 @@ def auc(x, y, reorder=False):
     """Compute average precision (AP) from prediction scores
 
-    This score corresponds to the area under the precision-recall curve.
+    Optionally, this will compute an eleven-point interpolated average 
```
trailing whitespaces somewhere

GaelVaroquaux commented:
I've addressed the comments and rewritten the example to start with a simple case.

The example can be seen here:
https://10944-843222-gh.circle-artifacts.com/0/home/ubuntu/scikit-learn/doc/_build/html/stable/auto_examples/model_selection/plot_precision_recall.html

GaelVaroquaux commented:
All green. Can I haz reviews / mrg ?

amueller commented on doc/whats_new.rst:
```diff
@@ -6,6 +6,35 @@ Release history
 
 Version 0.19
+==============
+
```
Still rebase issue

amueller commented on doc/whats_new.rst:
```diff
@@ -6,6 +6,35 @@ Release history
+     by the change in recall since the last operating point, as per the
+     `Wikipedia entry <http://en.wikipedia.org/wiki/Average_precision>`_.
+     (`#7356 <https://github.com/scikit-learn/scikit-learn/pull/7356>`_). By
+     `Nick Dingwall`_.
```
And @GaelVaroquaux ?

amueller commented on examples/model_selection/plot_precision_recall.py:
```diff
@@ -5,13 +5,18 @@
-low false negative rate. High scores for both show that the classifier is
-returning accurate results (high precision), as well as returning a majority of
-all positive results (high recall).
+Precision-Recall is a useful measure of success of prediction when the
```
Precision and recall? Or the precision-recall-curve?

vene replied:
yeah, and I would change the example title too: "Using precision and recall for classifier evaluation" or something. It's very awkward now.

And maybe replace below "very imbalanced" by "imbalanced"?

amueller commented on examples/model_selection/plot_precision_recall.py:
```diff
@@ -5,13 +5,18 @@
+relevant results are returned.
+
+The precision-recall curve shows the tradeoff between precision and
+recall for different threshold. A high area under the curve represents
```
decision thresholds?

vene replied:
high area -> large area?

amueller commented on examples/model_selection/plot_precision_recall.py:
```diff
@@ -71,76 +85,148 @@
+                                                    random_state=random_state)
+
+# Create a simple classifier
+classifier = svm.SVC(kernel='linear', probability=True,
```
Why????? How about ``LogisticRegression``? Or not using ``probability=True``? We don't need probabilities for the precision-recall-curve and many people are not aware of that. So maybe using ``SVC(linear)`` is actually good, but we shouldn't use ``probability=True``. What's the reason not to use ``LinearSVC()``?

amueller replied:
We're not even using ``predict_proba``, right?

GaelVaroquaux replied:
Aaaaaaaaaargh!

amueller commented on examples/model_selection/plot_precision_recall.py:
```diff
@@ -71,76 +85,148 @@
+# We use OneVsRestClassifier for multi-label prediction
+from sklearn.multiclass import OneVsRestClassifier
 # Run classifier
 classifier = OneVsRestClassifier(svm.SVC(kernel='linear', probability=True,
```
again, why ``probability=True``, why SVC(linear) instead of LinearSVC

GaelVaroquaux replied:
Yup!

amueller commented on examples/model_selection/plot_precision_recall.py:
```diff
@@ -152,23 +238,103 @@
+
+print("Target recall    Selected recall   Precision")
+for i in range(11):
+    print("  >= {}           {:3.3f}             {:3.3f}".format(i / 10,
```
Nitpick: if we're using format strings, maybe we should, you know, actually use them for the paddding?

amueller commented on examples/model_selection/plot_precision_recall.py:
```diff
@@ -152,23 +238,103 @@
+plt.fill_between(recall[iris_cls], precision[iris_cls], step='post', alpha=0.1,
+                 color='g')
+for i in range(11):
+    plt.annotate('',
```
If you put the '' on the next line you have more space ;)

GaelVaroquaux replied:
Yeah, but I don't feel it improves things.

amueller commented on examples/model_selection/plot_precision_recall.py:
```diff
@@ -54,11 +59,20 @@
+
+where :math:`P_n` and :math:`R_n` are the precision and recall at the
+nth threshold. A pair :math:`(R_k, P_k)` is referred to as an
+*operating point*.
```
Maybe mention that this is related to the integral under the curve but there are subtle differences?

amueller reviewed (commented):
Mostly nitpicks. I think it would be good to have an explanation in the tests and the whatsnew needs to be fixed, but I don't want to delay this because of minor issues in the example.

GaelVaroquaux commented:
Addressd all @amueller 's comments

GaelVaroquaux commented:
@amueller @vene : :+1: ?

vene commented on doc/whats_new.rst:
```diff
@@ -65,6 +65,14 @@ New features
+   - Added a `'eleven-point'` interpolated average precision option to
+     :func:`metrics.ranking.average_precision_score` as described in the
+     `PASCAL
+     Visual Object Classes (VOC) Challenge <http://citeseerx.ist.psu.edu/viewdoc/
```
I guess line wrapping here is weird?

vene commented on doc/whats_new.rst:
```diff
@@ -193,6 +201,13 @@ Enhancements
 .........
 
+   - :func:`metrics.ranking.average_precision_score` no longer linearly
+     interpolates between operating points, and instead weights precisions
```
weighs ?

vene commented on examples/model_selection/plot_precision_recall.py:
```diff
@@ -54,11 +59,20 @@
 
 Precision-recall curves are typically used in binary classification to study
-the output of a classifier. In order to extend Precision-recall curve and
+the output of a classifier. In order to extend the Precision-recall curve and
```
lowercase p

vene commented on examples/model_selection/plot_precision_recall.py:
```diff
@@ -71,76 +85,146 @@
+# In multi-label settings
+# ------------------------
+#
+# Create multli-label data, fit, and predict
```
multi

vene commented on examples/model_selection/plot_precision_recall.py:
```diff
@@ -71,76 +85,146 @@
+
+from sklearn.preprocessing import label_binarize
+
+# Use label_binarize to be multi-label like settings
```
wat :stuck_out_tongue:

GaelVaroquaux replied:
yeah :/. I don't like this part of the example, but that's what we have.

vene replied:
i just mean i don't understand what the comment is saying

vene commented on examples/model_selection/plot_precision_recall.py:
```diff
@@ -71,76 +85,146 @@
                                                     random_state=random_state)
 
+# We use OneVsRestClassifier for multi-label prediction
+from sklearn.multiclass import OneVsRestClassifier
```
newline after

vene commented on examples/model_selection/plot_precision_recall.py:
```diff
@@ -71,76 +85,146 @@
-# Compute Precision-Recall and plot curve
+
+###############################################################################
+# The precision-Recall score in multi-label settings
```
inconsistent capitalization. I'd suggest `precision-recall` everywhere.

I'm very confused now about what "the precision-recall score" is. Is it the same thing as average precision? I've never heard the former before. (just as two distinct scores.) The narrative at the top doesn't make it clear

vene commented on examples/model_selection/plot_precision_recall.py:
```diff
@@ -71,76 +85,146 @@
-plt.title('Precision-Recall example: AUC={0:0.2f}'.format(average_precision[0]))
-plt.legend(loc="lower left")
-plt.show()
+plt.title('Precision-Recall micro-averaged over all classes: AUC={0:0.2f}'
```
same confusion. Maybe "precision and recall micro-averaged..."?

GaelVaroquaux replied:
Or "micro average precision score"?

vene commented on examples/model_selection/plot_precision_recall.py:
```diff
@@ -152,23 +236,102 @@
+# ------------------------------
+#
+# In *interpolated* average precision, a set of desired recall values is
+# specified and for each desired value, we average the best precision
```
i'd move the comma to after the "and" on this line

vene commented on sklearn/metrics/ranking.py:
```diff
@@ -109,10 +105,14 @@ def auc(x, y, reorder=False):
 
-    This score corresponds to the area under the precision-recall curve.
+    Optionally, this will compute an eleven-point interpolated average
+    precision score: for each of the 11 evenly-spaced target recall values
```
maybe say eleven in words to make it easier to spot the connection to the interpolation arg below

vene commented on sklearn/metrics/ranking.py:
```diff
@@ -171,16 +191,48 @@ def average_precision_score(y_true, y_score, average="macro",
             y_true, y_score, sample_weight=sample_weight)
-        return auc(recall, precision)
+        # Return the step function integral
+        return -np.sum(np.diff(recall) * np.array(precision)[:-1])
```
I think this works because the last entry of precision is guaranteed to be 1, as written in the docstring of `precision_recall_curve`; do you think this warrants a note in a comment here, for posterity?

GaelVaroquaux replied:
Agreed

vene commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -123,6 +129,38 @@ def _average_precision(y_true, y_score):
+    precision = list(reversed(precision))
+    recall = list(reversed(recall))
+    indices = np.searchsorted(recall, np.arange(0, 1.1, 0.1))
+    return np.mean([max(precision[i:]) for i in indices])
```
confused: other than using lists instead of arrays, and other than **not ignoring** the first p-r pair with p=1.0, this seems identical to the actual implementation in this pr, what am I missing?

GaelVaroquaux replied:
Agreed.

Should we remove this test

vene commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -511,15 +552,18 @@ def test_precision_recall_curve_toydata():
-        assert_almost_equal(auc_prc, 0.25)
+        # Here we are doing a terrible prediction: we are always getting
+        # it wrong, hence the average_precision_score is the accuracy at
+        # change: 50%
```
chance

amueller commented:
LGTM, I haven't recalculated the test results though.

vene commented:
The "reference" implementation of the eleven-point average precision is not clear enough for my taste, so I wrote a simpler one, essentially:

```
precision, recall, _ = precision_recall_curve(y_true, y_score)

interpolated_ap = 0
recall_points = [0.0, .1, .2, .3, .4, .5, .6, .7, .8, .9, 1.0]
for recall_point in recall_points:
    # highest precision achieved for at least this much recall
    interpolated_ap += precision[recall >= recall_point].max()
interpolated_ap /= len(recall_points)
```

This implementation gets the same eleven-point interpolated score as the main code in this PR. Note that this reference implementation does **not** skip the terminal point at (p=1, r=0), a point that is not an operating point, but is returned by `precision_recall_curve`.

Because of this, the interpolated average precision at r=0 will always be 1 with this implementation. Based on  fig. 8.3 in [the ir book](https://nlp.stanford.edu/IR-book/html/htmledition/evaluation-of-ranked-retrieval-results-1.html) (the rotated one) I believe this should not be the case (otherwise that plot would start at 1).

GaelVaroquaux commented:
After discussion with @vene, I have removed the eleven-point average precision from this PR. Sorry @ndingwall we feel that get the fixes of this PR merged in is very important, but that we are not sure that we can warrant correctness the eleven point variant.

I will issue a new PR so that the work does not gets lost, and @ndingwall can take over.

vene commented on doc/whats_new.rst:
```diff
@@ -6,7 +6,7 @@ Release history
 
 Version 0.19
-============
+==============
```
what's up here?

GaelVaroquaux committed:
FIX: fix previous commit

GaelVaroquaux commented:
@vene: fixed

GaelVaroquaux commented:
@amueller and @vene gave me :+1: at the bar. Merging

GaelVaroquaux commented:
Merged

jnothman commented:
I think this needs more of a loud warning in changelog
